jupyter-org-client.el

```diff
@@ -416,6 +416,357 @@ the `syntax-table' will be set to that of the REPL buffers."
+Use a numeric prefix N to specify how many lines of context to use.
+Defaults to 3."
+  (interactive "p")
+  (let ((p '()))
+    (when (= 1 N) (setq N 3))
+    (save-excursion
+      (goto-char (point-min))
+      (while (re-search-forward org-babel-src-block-regexp nil t)
+        (push (list (format "line %s:\n%s"
+                            (line-number-at-pos (match-beginning 0))
+                            (save-excursion
+                              (goto-char (match-beginning 0))
+                              (let ((s (point)))
+                                (forward-line N)
+                                (buffer-substring s (point)))))
+                    (line-number-at-pos (match-beginning 0)))
+              p)))
+    (ivy-read "block: " (reverse p)
+              :action (lambda (candidate)
+                        (goto-char (point-min))
+                        (forward-line (1- (second candidate)))
+                        (outline-show-entry)
+                        (recenter)))))
+
+;;;###autoload
+(defun jupyter-org-edit-header ()
+  "Edit the src-block header in the minibuffer."
+  (interactive)
+  (let ((src-info (org-babel-get-src-block-info 'light)))
+    (unless src-info
+      (error "Not in a source block"))
+    (let* ((header-start (sixth src-info))
+           (header-end (save-excursion (goto-char header-start)
+                                       (line-end-position))))
+      (setf (buffer-substring header-start header-end)
+            (read-string "Header: "
+                         (buffer-substring header-start header-end))))))
+
+;;;###autoload
+(defun jupyter-org-kill-block-and-results ()
+  "Kill the block and its results."
+  (interactive)
+  (let ((src (org-element-context))
+        (result-start (org-babel-where-is-src-block-result))
+        end)
+    (if result-start
+        (save-excursion
+          (goto-char result-start)
+          (setq end (org-element-property :end src)))
+      (setq end (org-element-property :end src)))
+    (kill-region
+     (org-element-property :begin src)
+     end)))
+
+;;;###autoload
+(defun jupyter-org-copy-block-and-results ()
+  "Copy the src block at the current point and its results."
+  (interactive)
+  (let ((src (org-element-context))
+        (result-start (org-babel-where-is-src-block-result))
+        end)
+    (if result-start
+        (save-excursion
+          (goto-char result-start)
+          (setq end (org-babel-result-end)))
+      (setq end (org-element-property :end src)))
+    (kill-new
+     (buffer-substring
+      (org-element-property :begin src)
+      end))))
+
+;;;###autoload
+(defun jupyter-org-clone-block (&optional below)
+  "Clone the block."
+  (interactive "P")
+  (let* ((src (org-element-context))
+         (code (org-element-property :value src)))
+    (unless (eq (org-element-type src) 'src-block)
+      (error "Not in a source block"))
+    (jupyter-org-insert-src-block (not below))
```

Comment:
We can just use `(save-excursion (insert (org-element-interpret-data src)))` here right? After moving to the right location to insert the source block. This would remove the need for the rest of the lines below.

Reply:
I cannot quite use `org-element-interpret-data` or, rather, I prefer using `jupyter-org-insert-src-block` as then I don't have to go through the logic of figuring out the position where to clone a block. Finding the position is tricky because I need to check for results.

Reply:
I however saw that the logic to move to the next block was wrong and is not needed.